File: src/main/java/eu/mshade/mwork/event/EventBus.java
package eu.mshade.mwork.event;

import eu.mshade.mwork.ParameterContainer;

import java.util.List;
import java.util.Queue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.stream.Collectors;

public class EventBus<T> {

    private final Queue<EventContext<T>> eventContexts = new ConcurrentLinkedQueue<>();

    public <E extends T> EventContext<E> subscribe(Class<E> eventClass, EventListener<E> eventListener){
        DefaultEventContext<E> eventContext = new DefaultEventContext<>(eventClass, eventListener);
        eventContexts.add((DefaultEventContext<T>) eventContext);
        return eventContext;
    }

    public <E extends T> void publish(E event){
        this.publish(event, ParameterContainer.EMPTY);
    }

    public <E extends T> void publish(E event, ParameterContainer parameterContainer){
        List<EventContext<T>> eventContexts = this.eventContexts.stream()
                .filter(eventContext -> hasMatch(eventContext.getEventType(), (eventContext.getEventType().isInterface() && event.getClass().getInterfaces().length != 0 ? event.getClass().getInterfaces()[0] : event.getClass()), eventContext.getEventFilter()))
                .collect(Collectors.toList());
        if (!eventContexts.isEmpty()){
            for (EventPriority eventPriority : EventPriority.values()) {
                onEvent(event, parameterContainer, eventPriority, eventContexts);
            }
        }
    }

    public <E extends T> void publishAsync(E event){
        this.publishAsync(event, ParameterContainer.EMPTY);
    }

    public <E extends T> void publishAsync(E event, ParameterContainer parameterContainer) {
        CompletableFuture.runAsync(() -> publish(event, parameterContainer));
        //publish(event, parameterContainer);
    }


    private <E extends T> void onEvent(E event, ParameterContainer eventContainer, EventPriority eventPriority, List<EventContext<T>> eventContexts){
        //eventContexts.stream().filter(eventContext -> eventContext.getEventPriority() == eventPriority).forEach(eventContext -> scheduledExecutorService.execute(() -> eventContext.getEventListener().onEvent(event, eventContainer)));
        eventContexts.stream().filter(eventContext -> eventContext.getEventPriority() == eventPriority).forEach(eventContext -> eventContext.getEventListener().onEvent(event, eventContainer));
    }

    public <E extends T> List<EventContext<T>> getEventContexts(Class<E> eventType, EventFilter eventFilter){
        return eventContexts.stream().filter(eventContext -> hasMatch(eventContext.getEventType(), eventType, eventFilter)).collect(Collectors.toList());
    }

    private  boolean hasMatch(Class<?> from, Class<?> to, EventFilter eventFilter){
        return (eventFilter == EventFilter.ONLY ? from == to : from.isAssignableFrom(to));
    }

}

File: src/main/java/eu/mshade/mwork/event/EventBus.kt
package eu.mshade.mwork.event

import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.util.*

/**
 * The event bus manages listeners and allows you to trigger your own events.
 * @see [subscribe] to register a listener.
 * @see [publish] to trigger an event.
 * @author Luke and Oleksandr
 * @since 0.0.1
 */
class EventBus<T> {
    private val eventContexts: PriorityQueue<EventContext<*>> = PriorityQueue<EventContext<*>>(EventComparator())

    /**
     * Register a listener of an event.
     * @param eventClass the event class to listen to
     * @param eventListener your listener
     * @return An [EventContext] to manage listening options
     */
    fun <E : T> subscribe(eventClass: Class<E>, eventListener: EventListener<E>): EventContext<E> {
        val eventContext = DefaultEventContext(eventClass, eventListener)
        eventContexts.add(eventContext)
        return eventContext
    }

    /**
     * Allows you to unsubscribe any registered listener.
     * @param eventClass The event type of the listener
     * @param eventListener The listener to unsubscribe
     */
    fun <E : T> unsubscribe(eventClass: Class<E>, eventListener: EventListener<E>) {
        eventContexts.removeIf { it.eventType == eventClass && it.eventListener == eventListener }
    }

    /**
     * Pass an instance of an event to trigger all subscribed listeners.
     * @param event The event to trigger.
     */
    fun <E : T> publish(event: E) {
        eventContexts.stream()
            // Filter out contexts that don't match the event type or any subtype (EventFilter logic).
            .filter { eventContext -> hasMatch(
                eventContext.eventType.javaClass,
                if(eventContext.eventType.isInterface) eventContext.eventType.javaClass else
                    eventContext.eventType.interfaces[0].javaClass, eventContext.eventFilter
            )}
                // Trigger the listener on each context.
            .forEach { eventContext -> eventContext.eventListener.onEvent(event) }
    }

    /**
     * Do the same as [publish] but asynchronously.
     */
    fun <E : T> publishAsync(event: E) {
        runBlocking {
            launch {
                publish(event)
            }
        }
    }

    fun <E : T> getEventContexts(eventType: Class<E>, eventFilter: EventFilter): List<EventContext<T>> {
        return eventContexts.stream().filter { hasMatch(it.eventType, eventType, eventFilter) }
            .toList() as List<EventContext<T>>
    }

    /**
     * Check whether from and to are equals or if from is a subtype of to<br>
     * depending on the eventFilter.
     * @param from The event's class
     * @param to The event's class or any sub interface of it
     * @param eventFilter The event's filter
     */
    private fun hasMatch(from: Class<*>, to: Class<*>, eventFilter: EventFilter): Boolean {
        return if (eventFilter === EventFilter.ONLY) from == to else from.isAssignableFrom(to)
    }

    /**
     * A comparator to sort listeners by priority.
     */
    class EventComparator : Comparator<EventContext<*>> {
        override fun compare(o1: EventContext<*>, o2: EventContext<*>): Int {
            return o1.eventPriority.weight - o2.eventPriority.weight
        }

    }
}